cpp/src/interop/to_arrow_host.cu

```diff
@@ -0,0 +1,429 @@
+{
+  constexpr size_type BIT_WIDTH_RATIO = sizeof(__int128_t) / sizeof(DeviceType);
+  auto buf = std::make_unique<rmm::device_buffer>(input.size() * sizeof(__int128_t), stream, mr);
+
+  auto count = thrust::counting_iterator<size_type>(0);
+  thrust::for_each(rmm::exec_policy(stream, mr),
+                   count,
+                   count + input.size(),
+                   [in  = input.begin<DeviceType>(),
+                    out = reinterpret_cast<DeviceType*>(buf->data()),
+                    BIT_WIDTH_RATIO] __device__(auto in_idx) {
+                     auto const out_idx = in_idx * BIT_WIDTH_RATIO;
+                     // the lowest order bits are the value, the remainder
+                     // simply matches the sign bit to satisfy the two's
+                     // complement integer representation of negative numbers.
+                     out[out_idx] = in[in_idx];
+#pragma unroll BIT_WIDTH_RATIO - 1
+                     for (auto i = 1; i < BIT_WIDTH_RATIO; ++i) {
+                       out[out_idx + i] = in[in_idx] < 0 ? -1 : 0;
+                     }
+                   });
+
+  return buf;
+}
+
+template std::unique_ptr<rmm::device_buffer> decimals_to_arrow<int32_t>(
+  cudf::column_view input, rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr);
+
+template std::unique_ptr<rmm::device_buffer> decimals_to_arrow<int64_t>(
+  cudf::column_view input, rmm::cuda_stream_view stream, rmm::device_async_resource_ref mr);
+
+namespace {
+
+struct dispatch_to_arrow_host {
+  cudf::column_view column;
+  rmm::cuda_stream_view stream;
+  rmm::device_async_resource_ref mr;
+
+  int populate_validity_bitmap(ArrowBitmap* bitmap) const
+  {
+    if (!column.has_nulls()) { return NANOARROW_OK; }
+
+    NANOARROW_RETURN_NOT_OK(ArrowBitmapResize(bitmap, static_cast<int64_t>(column.size()), 0));
+    CUDF_CUDA_TRY(cudaMemcpyAsync(bitmap->buffer.data,
+                                  (column.offset() > 0)
+                                    ? cudf::detail::copy_bitmask(column, stream, mr).data()
+                                    : column.null_mask(),
+                                  bitmap->buffer.size_bytes,
+                                  cudaMemcpyDefault,
+                                  stream.value()));
+    return NANOARROW_OK;
+  }
+
+  template <typename T>
+  int populate_data_buffer(device_span<T const> input, ArrowBuffer* buffer) const
+  {
+    NANOARROW_RETURN_NOT_OK(ArrowBufferResize(buffer, input.size_bytes(), 1));
+    CUDF_CUDA_TRY(cudaMemcpyAsync(
+      buffer->data, input.data(), input.size_bytes(), cudaMemcpyDefault, stream.value()));
+    return NANOARROW_OK;
+  }
+
+  template <typename T,
+            CUDF_ENABLE_IF(!is_rep_layout_compatible<T>() &&
+                           (!std::is_same_v<T, numeric::decimal128> &&
+                            !std::is_same_v<T, numeric::decimal32> &&
+                            !std::is_same_v<T, numeric::decimal64>))>
```

Comment:
Similar to the dispatcher in `to_arrow_device.cu`

```suggestion
  template <typename T,
            CUDF_ENABLE_IF(!is_rep_layout_compatible<T>()>
```

Reply:
`to_arrow_device.cu` specifies an explicit `decimal128` specialization. that's different than this which is using `ENABLE_IF` to use the same impl for both decimal32 and decimal64 and letting decimal128 use the same implementation as the other fixed-width versions. 

the overloading / dispatch wouldn't work correctly if I remove the check here.